src/Data/Array/Mutable/Linear.hs

```diff
@@ -120,14 +120,15 @@ read = Unsafe.toLinear readUnsafe
     readUnsafe :: Array a -> Int -> (Array a, Unrestricted a)
     readUnsafe arr@(Array size mutArr) ix
       | indexInRange size ix =
-          let !val = Unsafe.readMutArr mutArr ix
-          in  (arr, Unrestricted val)
+          let !(# a #) = Unsafe.readMutArr mutArr ix
```

Comment:
I do not completely understand why we still need the bang pattern here, I thought pattern matching on the result would evaluate it even without the bang pattern. Without it, the read/write reordering issue reappears.

Reply:
That's actually fairly surprising to me too. It does sound wrong. A `let` at an unboxed value should always be evaluated. I could understand if `let (# a #) = …` was illegal syntax, I suppose (we can make a case that there shouldn't be a lazy pattern-matching on unboxed tuple). But I wouldn't have expected it to fail to evaluate its right hand side.

I did find the explanation, though, it's in the last paragraph of the [relevant manual section](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-UnboxedTuples). Note how the desugaring breaks the strictness assumption by adding an extra thunk around the computation.

It may be best to avoid `let (#…` and replace them with explicit `case` to avoid surprises. But `let !p = …` is, of course, equivalent to a `case`.

Reply:
To be honest, after I replaced the case with a `let` initially, I think the let binding looks clearer. However, I'm happy to replace it with an explicit `case` if that's what you prefer.

Reply:
It doesn't matter for today. Let's see how often we get owned by this, before deciding for a solution.